### PR TITLE
Merge plugin containers

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?
Adds functionality for merging plugins when they use a common image. If two plugins

1. Require the same sidecar image
2. Are compatible (i.e. merging them doesn't change anything in the spec)

then the broker will combine them into a single plugin with the plugin specs merged. If any issue or incompatibility is detected, the plugins are returned unmodified.

The PR is broken into several commits to make review easier:

- e756eb1 - Minor fix to `brokers/testdata/config-plugin-ids.json`, which contained references to plugins that have been removed from the registry
- 2ac7e1a - Add dependency on `k8s.io/apimachinery`; should be covered by CQ https://dev.eclipse.org/ipzilla/show_bug.cgi?id=17574
- ae9bb10 - Update vendor from adding above dependency
- c4b9659 - Implement the main merging functionality in `utils/mergeplugins`
- 5003ee0 - Add test cases for merging functionality
- 94c41f9 - Make artifacts and metadata broker use merging functionality
- 203f8f1 - Fixup docker images and CI; old version complained about using `%w` in `fmt.Errorf`
- c61c2d4 - Workaround to make Che Server accept merged plugins

**Note:** I had to include a workaround (commit f4b03e4), since the Che server will not accept any ChePlugins that don't match the devfile. To avoid this issue, the merged plugins will reuse the plugin ID for the first plugin in the list.

### What issues does this PR fix or reference?
Resolves https://github.com/eclipse/che/issues/15373

### How to test?
Images are built and pushed to
- `docker.io/amisevsk/che-plugin-metadata-broker:dev`
- `docker.io/amisevsk/che-plugin-artifacts-broker:dev`

To test merging, add e.g. both `redhat/java11/latest` and `redhat/quarkus-java11/latest` installed. For testing locally, you can change `brokers/testdata/config-plugin-ids.json` and run `make test-metadata` and `make test-artifacts`:

```json
[
  {
    "id": "redhat/quarkus-java11/latest"
  },
  {
    "id": "redhat/java11/latest"
  },
  {
    "id": "eclipse/che-theia/latest"
  }
]
```

For a full test, the following devfiles will trigger the merging functionality:
```
devfiles/apache-camel-k/devfile.yaml
devfiles/apache-camel-springboot/devfile.yaml
devfiles/dotnet-asp.net/devfile.yaml
devfiles/dotnet/devfile.yaml
devfiles/nodejs/devfile.yaml
devfiles/nodejs-mongo/devfile.yaml
devfiles/php-laravel/devfile.yaml
devfiles/php-mysql/devfile.yaml
devfiles/php-symfony/devfile.yaml
devfiles/php-web-simple/devfile.yaml
```

### CODE REVIEW

**Review by benoitf — COMMENTED** [review]
Hello, do we have a toggle ( like devfile property) to enable/disable this behavior ?

File: build/metadata/Dockerfile [hunk 1]
```diff
@@ -10,7 +10,7 @@
 #   Red Hat, Inc. - initial API and implementation
 #
 
-FROM golang:1.12-alpine3.9 as builder
+FROM golang:1.13-alpine3.12 as builder
```
Comment (sleshchenko) [inline build/metadata/Dockerfile:13]: Seems circleci needs to be updated as well https://app.circleci.com/pipelines/github/eclipse/che-plugin-broker/445/workflows/7469130a-3ff3-4796-baa8-a1b3981e171f/jobs/415

**Review by sleshchenko — APPROVED** [review]
Tested with provide make targets and seem to work as expected. Good job!

@benoitf Do you see the default behavior - merge plugins, and if there is any issue - users are able to disable it with devfile attribute. Or by default it's disabled, users are able to enable it and later when we're sure that it's stable - we just enable it by default?

File: build/metadata/Dockerfile (hunk 1 above)
Comment (sleshchenko) [reply build/metadata/Dockerfile:13]: Hope you don't mind that the fix is provided in https://github.com/eclipse/che-plugin-broker/pull/111/commits/f8d65b50dafc5fcd11790c827c23a68264d00fcb )

**Comment by benoitf** [comment]
@sleshchenko I would opt-in for now for the new behavior to see how it's behave

**Comment by benoitf** [comment]
but I'm fine if it's enabled by default with a flag to disable as well

I'm happy as long as we've a workaround to disable

Also maybe as we don't have many shared plugins we could just enable by default

File: utils/mergeplugins/merge.go [hunk 2]
```diff
@@ -0,0 +1,275 @@
+package mergeplugins
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply utils/mergeplugins/merge.go:1]: Woops

[the remark this answers is not in the document]
Comment (amisevsk) [reply utils/mergeplugins/merge.go:1]: Sounds like a potential CI action to add ;)

File: utils/mergeplugins/merge.go [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+	r.cpuRequest = resource.NewQuantity(0, resource.DecimalSI)
+	return r
+}
+
+// MergePlugins collapses a list of plugins by merging any plugins that share the same container into
+// a single plugin with multiple extensions
+func MergePlugins(plugins []model.PluginMeta) ([]model.PluginMeta, []string) {
+	var unmodified []model.PluginMeta
+	var logBuf []string
+	toMerge := map[string][]model.PluginMeta{}
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply utils/mergeplugins/merge.go:33]: I generally don't like numbers in variable names.

File: utils/mergeplugins/merge.go [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+	}
+	if !resources.cpuLimit.IsZero() {
+		container.CPULimit = resources.cpuLimit.String()
+	}
+	if !resources.cpuRequest.IsZero() {
+		container.CPURequest = resources.cpuRequest.String()
+	}
+}
+
+func getMergedPluginNames(plugins []model.PluginMeta) string {
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply utils/mergeplugins/merge.go:260]: Renamed to `formatPluginNames`

File: utils/mergeplugins/merge.go [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+		}
+	}
+	return nil
+}
+
+func addVolumes(container *model.Container, plugin model.PluginMeta) error {
+	for _, volume := range plugin.Spec.Containers[0].Volumes {
+		merge, fail := volumeMergable(volume, container.Volumes)
+		if fail {
+			return fmt.Errorf("volume %s collides with another in workspace", volume.Name)
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply utils/mergeplugins/merge.go:228]: Done

File: utils/mergeplugins/mergeable.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
+package mergeplugins
+
+import (
+	"reflect"
+	"strings"
+
+	"github.com/eclipse/che-plugin-broker/model"
+)
+
+func canBeMerged(plugin model.PluginMeta) bool {
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply utils/mergeplugins/mergeable.go:10]: Changed to `pluginMergeable`. Within the context of this package, though, I don't think there would be a collision.

File: utils/mergeplugins/merge.go [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+	return merged, nil
+}
+
+func addEnv(merged *model.PluginMeta, toMerge model.PluginMeta) error {
+
+	for _, env := range toMerge.Spec.WorkspaceEnv {
+		merge, fail := envMergable(env, merged.Spec.WorkspaceEnv)
+		if fail {
+			return fmt.Errorf("different values for environment variable %s", env.Name)
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply utils/mergeplugins/merge.go:136]: Good catch, I copy-pasted the function :)

File: utils/mergeplugins/merge_test.go [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+func TestMergePluginsForImageFailureModes(t *testing.T) {
+	tests := []struct {
+		name              string
+		args              []model.PluginMeta
+		expectedErrRegexp *regexp.Regexp
+	}{
+		{
+			name:              "Env cannot be merged",
+			args:              loadPluginMetasFromFile(t, "failure/conflicting_env_var.yaml").Metas,
+			expectedErrRegexp: regexp.MustCompile(`different values for environment variable`),
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply utils/mergeplugins/merge_test.go:48]: I think this is better solved by having test cases cover this, e.g. the test case only has a single env var for each meta and they conflict. Writing tests to match formatting exactly makes the tests more fragile (if every time you change the test data you have to change the tests, it just turns the test into a reminder that you're changing formatting).

File: utils/mergeplugins/merge.go [hunk 9]
```diff
@@ -63,15 +63,15 @@ func MergePlugins(plugins []model.PluginMeta) ([]model.PluginMeta, []string) {
 func mergePluginsForImage(image string, plugins []model.PluginMeta) (*model.PluginMeta, error) {
 	merged := &model.PluginMeta{
 		APIVersion: "v2",
-		Version:    "latest",
-		Name:       "merged-" + utils.SanitizeImage(image),
-		Publisher:  "eclipse-che",
+		Version:    plugins[0].Version,
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply utils/mergeplugins/merge.go:66]: Yeah that's partly a concern; we could work around it down the line by adding a field to `ChePlugin` that's something like
```
mergedPlugin: true
containsPlugins: []string{...}
```
and then propagate that information down.

This is an issue we'll run into with plugin dependencies (https://github.com/eclipse/che/issues/15376) as well. I've created issue https://github.com/eclipse/che/issues/17736 for tracking

File: build/metadata/Dockerfile (hunk 1 above)
Comment (amisevsk) [reply build/metadata/Dockerfile:13]: Thanks!

**Comment by amisevsk** [comment]
To add to the discussion of whether to enable by default or not, the following devfiles as they're currently written would trigger an attempt to merge:

```
devfiles/apache-camel-k/devfile.yaml
devfiles/apache-camel-springboot/devfile.yaml
devfiles/dotnet-asp.net/devfile.yaml
devfiles/dotnet/devfile.yaml
devfiles/nodejs/devfile.yaml
devfiles/nodejs-mongo/devfile.yaml
devfiles/nodejs-yarn/devfile.yaml
devfiles/nodejs-mongo/devfile.yaml
devfiles/php-laravel/devfile.yaml
devfiles/php-mysql/devfile.yaml
devfiles/php-symfony/devfile.yaml
devfiles/php-web-simple/devfile.yaml
```

File: utils/mergeplugins/merge_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply utils/mergeplugins/merge_test.go]: Added env var to regex to speed up review :)

**Comment by amisevsk** [comment]
Commit d537bef adds the flag `--merge-plugins` to enable merging. I'll file a PR to pass the flag on the Che server side once this one is merged.

**Review by sleshchenko — APPROVED** [review]

**Pull request merged**